### PR TITLE
TandemFoil: ANP cross-foil surface decoder (anp_srf=True)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The ANP (Attentive Neural Process) cross-foil surface decoder (`anp_srf=True`) is flagged as HIGH PRIORITY in the TandemFoilSet program and is part of the historical frontier lineage (#2319→#2350→#2357→#2379). It enables cross-foil information sharing between the upper and lower foil surfaces in tandem configuration. This is potentially the strongest single untested improvement on TandemFoilSet. We test it with and without the physics feature stack to understand its independent contribution.

## Current Baseline

- **val_primary/surface_pressure_mae: 262.82** (PR #2414, tanjiro — physics features + Lion lr=3e-4, 2 epochs)
- **test_primary/surface_pressure_mae: 257.51**

## Instructions

Run 4 parallel trials with `CUDA_VISIBLE_DEVICES`. Use `--wandb_group radford-tandem-anp` for all runs. Use AdamW lr=5e-4, model_slices=64, cosine_t_max=30.

Physics flags (used in Trials 0 and 1): `--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`

### Trial 0 — GPU 0: ANP decoder + physics features + AdamW (full combo)
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --anp-srf --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-anp --wandb_name shinji/tandem-anp-physics-adamw --agent shinji
```

### Trial 1 — GPU 1: ANP decoder + physics features + Lion (isolate ANP contribution vs AdamW)
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine_t_max 50 --model-slices 64 --anp-srf --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-anp --wandb_name shinji/tandem-anp-physics-lion --agent shinji
```

### Trial 2 — GPU 2: ANP decoder alone (no physics features, AdamW)
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --anp-srf --wandb_group radford-tandem-anp --wandb_name shinji/tandem-anp-only-adamw --agent shinji
```

### Trial 3 — GPU 3: No ANP, physics features + AdamW (matched control)
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-anp --wandb_name shinji/tandem-noanp-physics-adamw --agent shinji
```

Run all 4 in parallel (background each with `&`). Wait for all to finish.

**ANP implementation note:** The ANP head may slow training slightly (extra cross-attention). Note epoch time with vs without ANP.

**Critical things to report:**
- val_primary/surface_pressure_mae for each trial (best across all epochs)
- The val_re_rand split specifically — OOD generalization across Reynolds numbers is key paper metric
- Epochs completed per trial (note any slowdown from ANP head)
- Compare Trial 0 (ANP+physics+AdamW) vs Trial 3 (no ANP, physics+AdamW) to isolate ANP's contribution

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shinji:

## Results

### Summary Table

| Trial | Config | Epochs | val_primary (↓) | test_primary (↓) | val_re_rand (↓) | test_re_rand (↓) | Runtime |
|-------|--------|--------|-----------------|------------------|-----------------|------------------|---------|
| **Baseline** | Physics+Lion, 96 slices | **2** | **262.82** | **257.51** | — | **256.55** | ~30 min |
| 0 | ANP+physics+AdamW, 64 slices | 1 | 391.86 | 374.06 | 331.11 | 374.82 | 33 min |
| 1 | ANP+physics+Lion, 64 slices | 2 | 341.53 | 327.26 | 285.96 | 307.04 | 59 min |
| 2 | ANP only+AdamW, 64 slices | 2 | 338.15 | 320.32 | 269.86 | 296.18 | 57 min |
| 3 | No ANP+physics+AdamW, 64 slices | 1 | 384.14 | 368.19 | 330.19 | 360.10 | 32 min |

### Per-Split Test MAE (best trial = Trial 2)

| Split | Trial 0 | Trial 1 | Trial 2 | Trial 3 | Baseline |
|-------|---------|---------|---------|---------|----------|
| single_in_dist | 392.24 | 384.69 | 412.61 | 393.07 | 267.26 |
| geom_camber_rc | 402.50 | 349.33 | 293.55 | 422.27 | 280.59 |
| geom_camber_cruise | 326.67 | 268.00 | 278.92 | 297.31 | 225.63 |
| re_rand | 374.82 | 307.04 | 296.18 | 360.10 | 256.55 |

### W&B Run IDs

- Trial 0 (ANP+physics+AdamW): `yzo13zhg`
- Trial 1 (ANP+physics+Lion): `q5t8ekn3`
- Trial 2 (ANP only+AdamW): `0s3v72rq`
- Trial 3 (No ANP+physics+AdamW): `jwvlqps0`
- All runs in group: `radford-tandem-anp`

### Commands Used

```bash
# Trial 0
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max 30 --model-slices 64 --anp-srf --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-anp --wandb-name "shinji/tandem-anp-physics-adamw" --agent shinji

# Trial 1
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 50 --model-slices 64 --anp-srf --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-anp --wandb-name "shinji/tandem-anp-physics-lion" --agent shinji

# Trial 2
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max 30 --model-slices 64 --anp-srf --wandb-group radford-tandem-anp --wandb-name "shinji/tandem-anp-only-adamw" --agent shinji

# Trial 3
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max 30 --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-anp --wandb-name "shinji/tandem-noanp-physics-adamw" --agent shinji
```

### Peak Memory

Observed during training via nvidia-smi:
- Trial 0 (ANP+physics+AdamW): ~1 GB (completed quickly, 1 epoch)
- Trial 1 (ANP+physics+Lion): ~89 GB
- Trial 2 (ANP only+AdamW): ~68 GB
- Trial 3 (No ANP+physics+AdamW): ~1 GB (completed quickly, 1 epoch)

### What Happened

**None of the trials beat the baseline (262.82 val_primary).** The best trial was Trial 2 (ANP only, AdamW) at 338.15 — still 75 points worse.

**Key issue: `model_slices=64` was significantly slower than baseline's default 96 slices.** With 64 slices, epochs took ~30 min each, so Trials 0 and 3 (AdamW, cosine_t_max=30) only completed 1 epoch before the 30-min timeout. Trials 1 and 2 squeezed in 2 epochs because their first epoch finished just under 30 min (the timeout only checks at epoch boundaries).

**Isolating ANP's contribution (Trial 0 vs Trial 3, both 1 epoch):**
- Trial 0 (ANP+physics+AdamW): val=391.86, loss=0.902
- Trial 3 (No ANP, physics+AdamW): val=384.14, loss=0.962
- ANP slightly **hurt** val_primary (+7.7) despite achieving lower training loss. This suggests the ANP head's cross-attention may need more epochs to converge, or the surface correction signal needs longer to align with the main model predictions.

**Interesting signals:**
- Trial 2 (ANP only) achieved the best val_re_rand (269.86) — substantially better than other trials, suggesting ANP does help OOD generalization when given enough training.
- Trial 2 also had the best geom_camber_rc test MAE (293.55 vs baseline 280.59) — closest to baseline on this difficult split.
- Lion optimizer (Trial 1 vs Trial 0) helped substantially when given 2 epochs (341.53 vs 391.86), but still behind the baseline Lion configuration.

**Confounds making these results hard to interpret:**
1. `model_slices=64` vs baseline's 96 doubled epoch time, limiting effective training
2. The 1-epoch trials are not directly comparable to the 2-epoch baseline
3. The AdamW trials used cosine_t_max=30 (rapid LR cycling every 60 steps), which may not be optimal

### Suggested Follow-ups

1. **Re-run with default model_slices=96** to eliminate the slices confound and allow fair 2-epoch comparison against baseline.
2. **ANP + physics + Lion with 96 slices** — the most promising combo, matching baseline optimizer/slices but adding ANP.
3. **Longer training** — Trial 2's strong improvement from epoch 1 (480→338) suggests ANP benefits from more training. If epoch times can be reduced to ~15 min (with 96 slices), 2+ epochs would give a fairer test.
4. **ANP with cosine_t_max=50 or higher** — the rapid cycling with T_max=30 (step-based scheduler) creates extreme LR oscillation. Matching the baseline's T_max=50 would be fairer.
